### PR TITLE
Order unit partial refund

### PR DESCRIPTION
**Pull request by Zales0123**
Based on #47 

zrzut ekranu 2018-08-28 o 13 04 36

Everything is quite well prepared to introduce shipment partial refunds in next PR as well

### CODE REVIEW

File: src/Resources/config/doctrine/Refund.orm.xml [hunk 1]
```diff
@@ -8,10 +8,6 @@
 >
 
     <entity name="Sylius\RefundPlugin\Entity\Refund" table="sylius_refund_refund">
-        <unique-constraints>
-            <unique-constraint columns="refunded_unit_id,type" />
-        </unique-constraints>
```
Comment (Zales0123) [inline src/Resources/config/doctrine/Refund.orm.xml]: There is a missing migration for this page :(

File: features/having_credit_memo_generated.feature [hunk 2]
```diff
@@ -35,3 +35,12 @@ Feature: Having credit memo generated
         Then this credit memo should contain 1 "Mr. Meeseeks T-Shirt" product with "$0.90" tax applied
         And it should be issued in "United States" channel
         And its total should be "$9.90"
+
+    @ui @application
+    Scenario: Seeing the details of generated credit memo with partial price
+        Given 1st "Mr. Meeseeks T-Shirt" product from order "#00000022" has already "$5.50" refunded with "Space money" payment
+        And I browse the details of the only credit memo generated for order "#00000022"
```
Comment (GSadee) [inline features/having_credit_memo_generated.feature]: When

File: spec/Creator/RefundCreatorSpec.php [hunk 3]
```diff
@@ -62,4 +68,19 @@ function it_throws_exception_if_unit_has_already_been_refunded(
             ->during('__invoke', ['000222', 1, 1000, $refundType])
         ;
     }
+
+    function it_throws_exception_if_order_item_unit_refund_amount_it_too_big(
```
Comment (GSadee) [inline spec/Creator/RefundCreatorSpec.php]: ..._is_too_big

File: src/Resources/translations/messages.en.yml [hunk 4]
```diff
@@ -9,7 +9,9 @@ sylius_refund:
         issued_from: Issued from
         manage_credit_memos: Manage credit memos
         order_number: Order number
+        partial_refund: Partial refund
         refund_all: Refund all
+        refunded: Refunded
```
Comment (GSadee) [inline src/Resources/translations/messages.en.yml]: Wrong order of keys

File: tests/Behat/Context/Application/RefundingContext.php [hunk 5]
```diff
@@ -28,6 +30,9 @@
     /** @var RepositoryInterface */
     private $refundRepository;
 
+    /** @var RemainingTotalProviderInterface */
+    private $remainingOrderItemUnitTotalProvider;
```
Comment (pamil) [inline tests/Behat/Context/Application/RefundingContext.php]: `$remainingTotalProvider`?

File: src/Creator/RefundUnitsCommandCreator.php [hunk 6]
```diff
@@ -46,6 +48,13 @@ public function fromRequest(Request $request): Command
     private function parseIdsToUnitRefunds(array $units): array
     {
         return array_map(function (array $refundUnit): UnitRefund {
+            if (isset($refundUnit['amount']) && $refundUnit['amount'] !== '') {
+                $id = (int) $refundUnit['partial-id'];
```
Comment (pamil) [inline src/Creator/RefundUnitsCommandCreator.php]: I wonder whether using a form to create a command wouldn't be a better option - right now we have to use custom code for both command creation, form handling and form view rendering which can't be extended or reused as easily as forms (total would be `MoneyType` etc.).

File: src/Checker/UnitRefundingAvailabilityChecker.php [hunk 7]
```diff
@@ -6,22 +6,34 @@
...
 
     public function __invoke(int $unitId, RefundType $refundType): bool
     {
-        return null == $this
-            ->refundRepository
-            ->findOneBy(['refundedUnitId' => $unitId, 'type' => $refundType->__toString()])
-        ;
+        // temporary solution until providing the possibility to refund partial shipment as well
+        if ($refundType->equals(RefundType::shipment())) {
+            return null == $this
```
Comment (pamil) [inline src/Checker/UnitRefundingAvailabilityChecker.php]: Loose comparison alert! 🚨

**Comment by pamil** [comment]
Thanks Matt! 🎉

**Pull request merged**